commit c82b5a74cc739385db6e4275fe504a0e9469bf01
Author: Johannes Berg <johannes.berg@intel.com>
Date:   Sun Jul 14 23:39:20 2013 +0300

    mac80211: make active monitor injection work w/ HW queue
    
    When a driver (like hwsim) uses HW queue control an
    active monitor vif needs to be used for the queues,
    make the code do that. Otherwise we'd bail out and
    drop the frames.
    
    Signed-off-by: Johannes Berg <johannes.berg@intel.com>

diff --git a/net/mac80211/tx.c b/net/mac80211/tx.c
index f82301b6cef2..be4d3caf4879 100644
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -1272,6 +1272,10 @@ static bool __ieee80211_tx(struct ieee80211_local *local,
 
 	switch (sdata->vif.type) {
 	case NL80211_IFTYPE_MONITOR:
+		if (sdata->u.mntr_flags & MONITOR_FLAG_ACTIVE) {
+			vif = &sdata->vif;
+			break;
+		}
 		sdata = rcu_dereference(local->monitor_sdata);
 		if (sdata) {
 			vif = &sdata->vif;